python/cudf/cudf/dataframe/dataframe.py

```diff
@@ -1093,19 +1087,38 @@ def reset_index(self, drop=False):
 
     def take(self, positions, ignore_index=False):
         out = DataFrame()
+
         if self._cols:
             positions = columnops.as_column(positions).astype("int32").data.mem
             cols = [s._column for s in self._cols.values()]
             result_cols = cpp_copying.apply_gather(cols, positions)
             for i, col_name in enumerate(self._cols):
                 out[col_name] = result_cols[i]
 
-            if ignore_index:
-                out.index = RangeIndex(len(out))
-            else:
-                out.index = self.index.take(positions)
+            if isinstance(self.columns, cudf.MultiIndex):
+                out.columns = self.columns
+
+        if ignore_index:
+            out.index = RangeIndex(len(out))
+        elif len(out) == 0:
+            out = out.set_index(self.index.take(positions))
+        else:
+            out.index = self.index.take(positions)
+
         return out
 
+    def take_columns(self, positions):
```

Comment:
Do we want this in the public API? Pandas doesn't have a `take_columns`. Also curious: hat are the resulting column names if `self.columns` is a MultiIndex?

Reply:
You're right that this should not go into the public API - it is a helper method for multi-column indexing.  The column names become tuples and lose their `self.columns` as a MultiIndex until it is recomputed and assigned in the caller. Thanks!

Reply:
i took your advice